framework/src/org/checkerframework/framework/type/AnnotatedTypeMirror.java

```diff
@@ -925,29 +911,34 @@ public void setTypeArguments(List<? extends AnnotatedTypeMirror> ts) {
 
         /** @return the type argument for this type */
         public List<AnnotatedTypeMirror> getTypeArguments() {
-            if (typeArgs == null) {
-                typeArgs = new ArrayList<AnnotatedTypeMirror>();
-                if (wasRaw()) {
-                    // TODO: This doesn't handle recursive type parameter
-                    // e.g. class Pair<Y extends List<Y>> { ... }
-                    // Type argument inference for raw types can be improved. See Issue #635.
-                    // https://github.com/typetools/checker-framework/issues/635
-                    AnnotatedDeclaredType declaration =
-                            atypeFactory.fromElement((TypeElement) getUnderlyingType().asElement());
-                    for (AnnotatedTypeMirror typeParam : declaration.getTypeArguments()) {
-                        AnnotatedWildcardType wct =
-                                atypeFactory.getUninferredWildcardType(
-                                        (AnnotatedTypeVariable) typeParam);
-                        typeArgs.add(wct);
-                    }
-                } else if (!getUnderlyingType().getTypeArguments().isEmpty()) { // lazy init
-                    for (TypeMirror t : getUnderlyingType().getTypeArguments()) {
-                        typeArgs.add(createType(t, atypeFactory, declaration));
-                    }
+            if (typeArgs != null) {
+                return typeArgs;
+            } else if (wasRaw()) {
+                // Initialize the type arguments with uninferred wildcards.
+                BoundsInitializer.initializeTypeArgs(this);
+
+                // Copy annotations from the declaration to the wildcards.
+                AnnotatedDeclaredType declaration =
+                        atypeFactory.fromElement((TypeElement) getUnderlyingType().asElement());
+                for (int i = 0; i < typeArgs.size(); i++) {
+                    AnnotatedTypeVariable typeParam =
+                            (AnnotatedTypeVariable) declaration.getTypeArguments().get(i);
+                    AnnotatedWildcardType wct = (AnnotatedWildcardType) typeArgs.get(i);
+                    AnnotatedTypeMerger.merge(typeParam.getUpperBound(), wct.getExtendsBound());
+                    wct.getSuperBound()
+                            .replaceAnnotations(typeParam.getLowerBound().getAnnotations());
+                    wct.replaceAnnotations(typeParam.getAnnotations());
                 }
                 typeArgs = Collections.unmodifiableList(typeArgs);
+                return typeArgs;
+            } else if (getUnderlyingType().getTypeArguments().isEmpty()) {
+                typeArgs = Collections.emptyList();
+                return typeArgs;
+            } else {
+                BoundsInitializer.initializeTypeArgs(this);
+                typeArgs = Collections.unmodifiableList(typeArgs);
```

Comment:
You said that `initializeTypeArgs` doesn't add annotations to the types. Why doesn't this branch need to add annotations?

Reply:
This type might have explicit annotations on the type arguments or the annotations are added via defaulting or some other process.  (Note, adding annotations to raw types in this method is the previous behavior.)